# Reading Exercise

## Imports

#### Libraries

In [25]:
# Library Imports
import pandas as pd
import numpy as np
import os

#### DF Imports

In [26]:
path = r'C:\Users\Chase\anaconda_projects\Exercise 4\07-2025 Instacart Basket Analysis'

In [27]:
# Original - Products csv file pathway
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [28]:
# Prepared - Orders WRANGLED csv file pathway
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = None)

#### Start

In [29]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### Test Dataframe (Mixed-type columns)

In [30]:
# Create a dataframe called: df_test
df_test = pd.DataFrame()

In [31]:
# Create a mixed type column within df_test
df_test['mix'] = ['a', 'b', 1, True]

In [32]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [33]:
# Check for mix-types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Chase\AppData\Local\Temp\ipykernel_39832\3698035231.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [34]:
df_test['mix'] = df_test['mix'].astype('str')

#### Missing Values

In [35]:
# Missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [36]:
# Create subset of dataframe containing missing values
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [37]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [38]:
df_prods.shape

(49693, 5)

In [39]:
# New dataframe clean - using False instead of True because you want non-missing values in the new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [40]:
# compare the rows 
df_prods_clean.shape

(49677, 5)

In [41]:
# Another way you can drop all missing values is via:    df_prods.dropna(inplace = True)

In [42]:
# If you wanted to drop NaNs from a particular column:   df_prods.dropna(subset = [‘product_name’], inplace = True)

#### Finding Duplicates

In [43]:
# Find full duplicates within your dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [44]:
# Display duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

#### Addressing Duplicates

In [45]:
# For comparision
df_prods_clean.shape

(49677, 5)

In [46]:
# New dataframe that doesn't include duplicates (Delete Duplicates)
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [47]:
# Compare and check work
df_prods_clean_no_dups.shape

(49672, 5)

In [48]:
nnnnnnnnnnnndf_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'), index=False)

NameError: name 'nnnnnnnnnnnndf_prods_clean_no_dups' is not defined

# ----Exercise 4.5 - df_ords Data Consistency Checks----

In [49]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [50]:
df_ords.shape

(3421083, 6)

#### Question 2

'unanmed:0': Is suspicious to me, what is this column supposed to represent?
'order_number': MAX order number is capped at 100 which seems weird. 
'orders_day_of_week': The weekday is 1-7 unless it's counting 0-6 as the full week.
'order_hour_of_day': Hours of the day are 1-24hrs but here they have 0-23hrs instead 
'days_since_prior_order': count doesn't match the other columns indicating possible missing values, MIN/MAX is 0-30 days odd because months cap at 31 days at the MAX.


#### Question 3

In [51]:
for col in df_ords.columns.tolist():
    weird = (df_ords[col].map(type) != df_ords[col].iloc[0].__class__).any()
    if weird:
        print(col)

order_id
user_id
order_number
orders_day_of_week
order_hour_of_day
days_since_prior_order


In [52]:
print(weird)

True


In [53]:
df_ords.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 6 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   order_number            int64  
 3   orders_day_of_week      int64  
 4   order_hour_of_day       int64  
 5   days_since_prior_order  float64
dtypes: float64(1), int64(5)
memory usage: 156.6 MB


#### Question 4

In [54]:
# Worked with Copilot to ensure that there weren't any false/positives within days_since_prior_order 
type(df_ords[col].iloc[0])

numpy.float64

In [55]:
# verify if values are all floats
df_ords[col].apply(type).value_counts()

days_since_prior_order
<class 'float'>    3421083
Name: count, dtype: int64

#### Question 5

In [56]:
# Missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

This shows there are 206,209 missing values in 'days_since_prior_order'.

#### Question 6

In [57]:
# Create a subset to isolate the missing values
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() ==True]

In [58]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [59]:
# Transform NaN into a numeric value of 0
df_ords_nan['days_since_prior_order'].fillna(0, inplace=True)

C:\Users\Chase\AppData\Local\Temp\ipykernel_39832\3565186252.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords_nan['days_since_prior_order'].fillna(0, inplace=True)
C:\Users\Chase\AppData\Local\Temp\ipykernel_39832\3565186252.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_ords_nan['days_since_prior_order'].fillna(0, inplace=T

In [60]:
# Check work
df_ords_nan.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [61]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     0.0  
11                      11                     0.0  
26                      14                     0.0  
39                      11                     0.0  
45                      12                     0.0  
...                    ...                     ...  
3420930                 12                     0.0  
3420934                 18                     0.0  
3421002                 19                     0.0  
3421019                 15                     0.0  
3421069                 11                     0.0  

[206209 rows x 6 columns]

In 'days_since_prior_order' I seleceted to make the NaN = 0 because there wasn't any data to suggest they were actual errors or missin values and the missing values seemed intentional

#### Question 7

In [62]:
# Find full duplicates within your dataframe
df_ords_dups = df_ords[df_ords.duplicated()]

In [63]:
# Display duplicates
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

No duplicates found

#### Question 8

No duplicates to fix

#### Question 9

In [64]:
df_ords_nan.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index=False)

NameError: name 'nnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnnndf_ords_nan' is not defined

# df_ords Dataset Fix

In [58]:
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

This is saying there are 206,209 rows in days_since_prior_order with NaNs. I accidently exported this subset which is what threw everything off for my assignments. I fixed this below.

In [69]:
# Fill NaNs with 0 — makes sense because there's no prior order
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].fillna(0)

In [70]:
# Create a flag for first orders
df_ords['first_order'] = df_ords['order_number'] == 1

In [71]:
# data check
df_ords['days_since_prior_order'].isna().sum()

np.int64(0)

In [73]:
# Confirm 1st order flag matches OG NaNs
df_ords['first_order'].sum()

np.int64(206209)

In [74]:
df_ords.shape

(3421083, 7)

### Export Clean

In [75]:
# Exporting the cleaned orders dataframe
# Includes all 3.4M rows with missing values in 'days_since_prior_order' imputed
# Also includes 'first_order' flag for semantic clarity
# Avoided exporting df_ords_nan, which only contains the subset of first orders
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'), index=False)